json/test_yajl.c

```diff
@@ -0,0 +1,184 @@
+#include <libnotify.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <yajl/yajl_parse.h>
+#include <yajl/yajl_version.h>
+
+#if (YAJL_MAJOR != 2)
+#error "yajl 2.x is required"
+#endif
+
+#ifdef __clang__
+#define COMPILER "clang"
+#else
+#define COMPILER "gcc"
+#endif
+
+
+struct coordinate_s {
+	double x;
+	double y;
+	double z;
+};
+
+typedef struct coordinate_s coordinate_t;
+
+struct context_s {
+	size_t depth;
+	size_t len;
+	char item;
+	char in_coord : 1;
+
+	coordinate_t c;
+};
+
+typedef struct context_s context_t;
+
+int cb_double(void* ctx_v, double value) {
+	context_t* ctx = ctx_v;
+
+	switch (ctx->item) {
+		case 'x': ctx->c.x += value; break;
+		case 'y': ctx->c.y += value; break;
+		case 'z': ctx->c.z += value; break;
+		default:;
+	}
+
+	return 1;
+}        
+
+int cb_start_map(void* ctx_v) {
+	context_t* ctx = ctx_v;
+
+	ctx->depth++;
+
+	if (ctx->depth == 2 && ctx->in_coord) {
+		ctx->len++;
+	}
+
+	return 1;
+}
+
+int cb_map_key(void* ctx_v, const unsigned char* key, size_t length) {
+	context_t* ctx = ctx_v;
+
+	if (ctx->depth == 1) {
+		if (strncmp("coordinates", (const char*)key, length) == 0) {
+			ctx->in_coord = 1;
+		} else {
+			ctx->in_coord = 0;
+		}
+	} else if (ctx->depth == 2 && ctx->in_coord) {
+		if (strncmp("x", (const char*)key, length) == 0) {
+			ctx->item = 'x';
+		} else if (strncmp("y", (const char*)key, length) == 0) {
+			ctx->item = 'y';
+		} else if (strncmp("z", (const char*)key, length) == 0) {
+			ctx->item = 'z';
+		} else {
+			ctx->item = 0;
+		}
+	}
+
+	return 1;
+}
+
+int cb_end_map(void* ctx_v) {
+	context_t* ctx = ctx_v;
+
+	ctx->depth--;
+
+	return 1;
+}
+
+int cb_end_array(void* ctx_v) {
+	context_t* ctx = ctx_v;
+
+	if (ctx->depth == 1 && ctx->in_coord) {
+		ctx->c.x /= ctx->len;
+		ctx->c.y /= ctx->len;
+		ctx->c.z /= ctx->len;
+	}
+
+	return 1;
+}
+
+static yajl_callbacks ycallbacks = {
+	NULL, /* null */
+	NULL, /* boolean */
+	NULL, /* integer */
+	&cb_double,
+	NULL, /* number */
+	NULL, /* string */
+	&cb_start_map,
+	&cb_map_key,
+	&cb_end_map,
+	NULL, /* start_array */
+	&cb_end_array
+};
+
+#define EMPTY_CTX ((context_t){0, 0, 0, 0, {0.0, 0.0, 0.0}})
+
+int main() {
+	const char test_buffer_1[] = "{\"coordinates\":[{\"x\":2.0,\"y\":0.5,\"z\":0.25}]}";
+	const char test_buffer_2[] = "{\"coordinates\":[{\"y\":0.5,\"x\":2.0,\"z\":0.25}]}";
+
+	context_t ctx = EMPTY_CTX;
+	yajl_handle yajl = yajl_alloc(&ycallbacks, NULL, &ctx);
+
+	yajl_parse(yajl, (const unsigned char*)test_buffer_1, sizeof(test_buffer_1) / sizeof(test_buffer_1[0]));
+	yajl_complete_parse(yajl);
+
+	if (!(ctx.c.x == 2.0 && ctx.c.y == 0.5 && ctx.c.z == 0.25)) {
```

Comment:
Please declare another const context to compare against, and use its values, e.g.

```c
const context_t expected = {
  .c = {
    .x = 2.0,
   ...
  }
};
...
if (ctx.c.c != expected.c.y || ...)
```